src/request.rs

```diff
@@ -1,46 +1,102 @@
 //! creating and sending HTTP requests
 use crate::{
     error,
-    response::{Headers, Response, CR_LF_2},
+    response::{find_slice, Headers, Response, CR_LF_2},
     tls,
     uri::Uri,
 };
 use std::{
     fmt,
-    io::{self, Read, Write},
+    io::{self, ErrorKind, Read, Write},
     net::{TcpStream, ToSocketAddrs},
     path::Path,
-    time::Duration,
+    time::{Duration, Instant},
 };
 
 const CR_LF: &str = "\r\n";
+const BUF_SIZE: usize = 8 * 1024;
+const SMALL_BUF_SIZE: usize = 8 * 10;
+const TEST_FREQ: usize = 200;
+const CONNECTION_TIMEOUT: u64 = 3 * 3600;
 
-///Copies data from `reader` to `writer` until the specified `val`ue is reached.
+///Copies data from `reader` to `writer` until the `timeout` is reached.
 ///Returns how many bytes has been read.
-pub fn copy_until<R, W>(reader: &mut R, writer: &mut W, val: &[u8]) -> Result<usize, io::Error>
+pub fn copy_with_timeout<R, W>(reader: &mut R, writer: &mut W, timeout: Instant) -> io::Result<u64>
 where
     R: Read + ?Sized,
     W: Write + ?Sized,
 {
-    let mut buf = Vec::with_capacity(200);
+    let mut buf = [0; BUF_SIZE];
+    let mut copied = 0;
+    let mut counter = 0;
+
+    loop {
+        let len = match reader.read(&mut buf) {
```

Comment:
I'm not 100% sure, but I expect this call would block indefinitely if it's waiting on more data to arrive, defeating the point of the deadline.

I believe you can get around this by setting read timeout to `initial_request_time + timeout - current_time` every once in a while, see e.g. https://github.com/neonmoe/minreq/commit/41ef16e33501712e64a5711a7f4bc132d9bfbe40

Reply:
I need to take a closer look at this.